core/src/main/java/org/apache/iceberg/util/BinPacking.java

```diff
@@ -169,19 +179,21 @@ public boolean hasNext() {
 
   private static class Bin<T> {
     private final long targetWeight;
+    private final int itemsPerBin;
     private final List<T> items = Lists.newArrayList();
     private long binWeight = 0L;
 
-    Bin(long targetWeight) {
+    Bin(long targetWeight, int itemsPerBin) {
       this.targetWeight = targetWeight;
+      this.itemsPerBin = itemsPerBin;
     }
 
     List<T> items() {
       return items;
     }
 
     boolean canAdd(long weight) {
-      return binWeight + weight <= targetWeight;
+      return binWeight + weight <= targetWeight && items.size() <= itemsPerBin;
```

Comment:
I think there is a slight issue with this implementation and how it interacts with lookback.

Consider we have a lookback of 3 a max number of items per bin of 2 and a max value of bin of 10
```
// Add Item (6)
[ 8 ], [3, 4], [4, 3]
// Even if remove largest is set largest bin is [8] but we actually have 2 bins we know we can never add another item to which we should be dropping first.

I think to properly do this we need to also finish off any bins which have reached their max item size
```

Reply:
OK, I will push a new commit to do this with UT case